### PR TITLE
Remove the varargs-based mdarray helper.

### LINKED ISSUE #62747: Constructing MD arrays in R2R code does not work on OSX/ARM64
https://github.com/dotnet/runtime/pull/61938 exposed this issue. 

We had other issues that could cause related test scenario (`TestGenericMDArrayBehavior`) to never run in R2R mode on OSX.
 #61938 makes R2R enabled more reliably on OSX, but now we see the test failing, so I disabled the scenario as a part of #61938.

I beleive the root cause is variadic JIT helper - `HCIMPL2VA(Object*, JIT_NewMDArr, CORINFO_CLASS_HANDLE classHnd, unsigned dwNumArgs) `. 
It looks like OSX/ARM64 has special ABI for variadic calls, but we call it just as a cdecl with extra arguments. As a result the helper reads garbage and allocating MD arrays fails.

We do not use this helper in regular JIT code. There is a comment that it may be worth switching R2R to non-variadic helper as well since it would be more portable. Perhaps it is time?

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Rename the non-varargs helper to the old name of the varargs helper as it has a better name.

Remove some now-unused varargs-specific macros.

Fixes #62747 

Reviewing without whitespace changes is recommended for the changes in the JIT.

### CODE REVIEW

**Comment by VSadov** [comment]
I think the change should also re-enable `TestGenericMDArrayBehavior` test on OSX

**Comment by jkoritzinsky** [comment]
Linux_musl ARM test failure is https://github.com/dotnet/runtime/issues/62873 (I made the criterion for the tracking issue too fine-grained and missed the musl instances of the failure)

**Review by VSadov — APPROVED** [review]
LGTM Thanks!!!

File: docs/design/coreclr/botr/readytorun-format.md [hunk 1]
```diff
@@ -719,81 +719,80 @@ enum ReadyToRunHelper
...
-    READYTORUN_HELPER_NewObject                 = 0x60,
-    READYTORUN_HELPER_NewArray                  = 0x61,
-    READYTORUN_HELPER_CheckCastAny              = 0x62,
-    READYTORUN_HELPER_CheckInstanceAny          = 0x63,
-    READYTORUN_HELPER_GenericGcStaticBase       = 0x64,
-    READYTORUN_HELPER_GenericNonGcStaticBase    = 0x65,
-    READYTORUN_HELPER_GenericGcTlsBase          = 0x66,
-    READYTORUN_HELPER_GenericNonGcTlsBase       = 0x67,
-    READYTORUN_HELPER_VirtualFuncPtr            = 0x68,
+    READYTORUN_HELPER_NewObject                 = 0x5F,
```
Comment (jkotas) [inline docs/design/coreclr/botr/readytorun-format.md:724]: There are intentional holes in these ranges so that it is possible to add a new helper ID in its most logical place, without renumbering (and thus breaking) all helper ID.

Could you please revert this change?

Comment (jkotas) [reply docs/design/coreclr/botr/readytorun-format.md:724]: Also, I think we want to maintain compatibility for `ILCompiler.Reflection.ReadyToRun` (ie make it possible to use it to dump older R2R version) and re-numbering all helpers makes it hard.

Comment (jkoritzinsky) [reply docs/design/coreclr/botr/readytorun-format.md:724]: I’ll undo the renumbering and leave the hole.

**Review by jakobbotsch — APPROVED** [review]

**Review by jkotas — APPROVED** [review]
Thanks!
